### PR TITLE
Don't delete MemoryPlanningLine referenced in outputs

### PR DESCRIPTION
**Pull request by voznesenskym**

### CODE REVIEW

File: torchinductor/codegen/wrapper.py [hunk 1]
```diff
@@ -296,10 +296,24 @@ def generate(self):
         result = IndentedBuffer()
         result.splice(self.header)
         result.splice(self.prefix)
+
+        def extract_name(out):
```
Comment (voznesenskym) [inline torchinductor/codegen/wrapper.py:300]: this feels gross, I wonder if it would be better served as a util on ir.py?

**Comment by voznesenskym** [comment]
Benchmarking this now

**Comment by voznesenskym** [comment]
Bench seems fine.

File: torchinductor/codegen/wrapper.py [hunk 2]
```diff
@@ -296,10 +296,24 @@ def generate(self):
...
+            o = out
+            while hasattr(o, "data"):
+                if hasattr(o, "name"):
+                    return o.name
+                else:
+                    o = o.data
+            return ""
+
+        out_names = [extract_name(out) for out in V.graph.graph_outputs]
```
Comment (jansel) [inline torchinductor/codegen/wrapper.py:309]: We collect names of all outputs here:
https://github.com/pytorch/torchdynamo/blob/da8bb9a875beede7b721425087d06ebae7bc76f8/torchinductor/scheduler.py#L849
and, here:
https://github.com/pytorch/torchdynamo/blob/da8bb9a875beede7b721425087d06ebae7bc76f8/torchinductor/scheduler.py#L600

I'd suggest adding a function like:
```
V.graph.get_output_names()
````

That combines the 3 places

File: torchinductor/codegen/wrapper.py [hunk 3]
```diff
@@ -296,10 +296,24 @@ def generate(self):
...
         with result.indent():
             while self.lines and isinstance(self.lines[-1], MemoryPlanningLine):
                 # these lines will be pointless
-                self.lines.pop()
+                if self.lines[-1].node.name not in out_names:
```
Comment (jansel) [inline torchinductor/codegen/wrapper.py:313]: Just put this in the condition of the while loop above to be more concise

**Review by jansel — CHANGES_REQUESTED** [review]

File: torchinductor/graph.py [hunk 4]
```diff
@@ -309,3 +308,10 @@ def compile_to_fn(self):
             setattr(mod, name, value)
 
         return mod.call
+
+    def get_output_names(self):
```
Comment (jansel) [inline torchinductor/graph.py:312]: Update the other uses to use this

**Review by jansel — APPROVED** [review]

**Pull request merged**